File: ios/GaelGrounds/Models/UserData.swift
import Foundation

struct UserProfile: Codable, Identifiable, Hashable {
    let id: UUID
    var displayName: String?
    var avatarUrl: String?
    var supportedCountyId: UUID?
    var isPremium: Bool = false
    var premiumExpiresAt: Date?
    var bestMatchId: UUID?
    // Explicit, opt-in consent to appear on the public Leaderboard (name +
    // stats visible to every other signed-in user) -- defaults to false on
    // every row, including existing premium users, per App Store guideline
    // 5.1.2: premium status alone must never be treated as consent to
    // publish someone's data. See LeaderboardView.swift's entries filter,
    // which requires both isPremium and this before including someone.
    var leaderboardOptIn: Bool = false
    let createdAt: Date?
}

struct UserProfilePremiumUpdate: Encodable {
    let isPremium: Bool
    let premiumExpiresAt: Date?
}

struct UserProfileUpdate: Encodable {
    let displayName: String
}

struct UserProfileAvatarUpdate: Encodable {
    let avatarUrl: String
}

struct UserProfileBestMatchUpdate: Encodable {
    let bestMatchId: UUID?
}

struct UserProfileLeaderboardOptInUpdate: Encodable {
    let leaderboardOptIn: Bool
}

struct SupportedCountyUpdate: Encodable {
    let supportedCountyId: UUID
}

struct UserVisit: Codable, Identifiable, Hashable {
    let id: UUID
    let userId: UUID
    let groundId: UUID
    let visitedAt: Date
    let notes: String?
    let photoUrls: [String]
    let createdAt: Date?
}

struct UserVisitInsert: Encodable {
    let groundId: UUID
    let userId: UUID
    let notes: String?
    let photoUrls: [String]?

    init(groundId: UUID, userId: UUID, notes: String?, photoUrls: [String]? = nil) {
        self.groundId = groundId
        self.userId = userId
        self.notes = notes
        self.photoUrls = photoUrls
    }
}

struct UserVisitPhotoUpdate: Encodable {
    let photoUrls: [String]
}

struct UserMatchAttendance: Codable, Identifiable, Hashable {
    let id: UUID
    let userId: UUID
    let matchId: UUID
    let notes: String?
    let photoUrls: [String]
    let createdAt: Date?
}

struct UserMatchAttendanceInsert: Encodable {
    let matchId: UUID
    let userId: UUID
}

struct AchievementDefinition: Codable, Identifiable, Hashable {
    struct RuleParams: Codable, Hashable {
        let count: Int?
        let countyId: UUID?
        let sportCode: SportCode?
        let province: Province?
    }

    let id: UUID
    let code: String
    let title: String
    let description: String
    let icon: String?
    let ruleType: String
    let ruleParams: RuleParams
    let createdAt: Date?
}

struct HomeAchievementKey: Hashable {
    let countyId: UUID
    let sportCode: SportCode
}

struct AchievementUnlock: Identifiable, Hashable {
    let id: UUID
    let title: String
    let description: String
    let icon: String?
    let tier: AchievementTier?
}

enum AchievementTier: String, Hashable, CaseIterable {
    case standard
    case bronze
    case silver
    case gold

    static func forHomeMatchCount(_ count: Int) -> AchievementTier {
        if count >= 50 { return .gold }
        if count >= 25 { return .silver }
        if count >= 10 { return .bronze }
        return .standard
    }

    var label: String {
        switch self {
        case .standard: return ""
        case .bronze: return "Bronze"
        case .silver: return "Silver"
        case .gold: return "Gold"
        }
    }
}

struct AchievementProgress: Hashable {
    let title: String
    let message: String
    let icon: String?
    let tier: AchievementTier
    let homeGameCount: Int
}

struct AchievementEvaluation: Hashable {
    let unlocks: [AchievementUnlock]
    let progress: AchievementProgress?
}

struct UserAchievement: Codable, Identifiable, Hashable {
    let id: UUID
    let userId: UUID
    let achievementId: UUID
    let unlockedAt: Date
    var pinned: Bool = false
}

struct UserAchievementInsert: Encodable {
    let achievementId: UUID
    let userId: UUID
}

struct UserAchievementPinnedUpdate: Encodable {
    let pinned: Bool
}

struct Friendship: Codable, Identifiable {
    let id: UUID
    let requesterId: UUID
    let addresseeId: UUID
    let status: String
    let createdAt: Date?
}

struct FriendshipInsert: Encodable {
    let requesterId: UUID
    let addresseeId: UUID
}

struct FriendshipStatusUpdate: Encodable {
    let status: String
}

struct DevicePushTokenUpsert: Encodable {
    let userId: UUID
    let token: String
    let updatedAt: Date
}

struct MatchReport: Codable, Identifiable {
    let id: UUID
    let matchId: UUID
    let userId: UUID
    let issueTypes: [String]
    let details: String?
    let status: String
    let createdAt: Date?
}

struct MatchReportInsert: Encodable {
    let matchId: UUID
    let userId: UUID
    let issueTypes: [String]
    let details: String?
}

File: android/app/src/main/kotlin/ie/gaelgrounds/app/data/model/UserData.kt
package ie.gaelgrounds.app.data.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

// Mirrors the update payload structs in ios/GaelGrounds/Models/UserData.swift --
// small Encodable shapes used for targeted .update() calls rather than
// resending the whole row.

@Serializable
data class UserProfilePremiumUpdate(
    @SerialName("is_premium") val isPremium: Boolean,
    @SerialName("premium_expires_at") val premiumExpiresAt: String?,
)

@Serializable
data class UserProfileUpdate(
    @SerialName("display_name") val displayName: String,
)

@Serializable
data class UserProfileAvatarUpdate(
    @SerialName("avatar_url") val avatarUrl: String,
)

@Serializable
data class UserProfileBestMatchUpdate(
    @SerialName("best_match_id") val bestMatchId: String?,
)

@Serializable
data class SupportedCountyUpdate(
    @SerialName("supported_county_id") val supportedCountyId: String,
)

@Serializable
data class UserVisit(
    val id: String,
    @SerialName("user_id") val userId: String,
    @SerialName("ground_id") val groundId: String,
    @SerialName("visited_at") val visitedAt: String,
    val notes: String? = null,
    @SerialName("photo_urls") val photoUrls: List<String> = emptyList(),
    @SerialName("created_at") val createdAt: String? = null,
)

@Serializable
data class UserVisitInsert(
    @SerialName("ground_id") val groundId: String,
    @SerialName("user_id") val userId: String,
    val notes: String? = null,
    @SerialName("photo_urls") val photoUrls: List<String>? = null,
)

@Serializable
data class UserVisitPhotoUpdate(
    @SerialName("photo_urls") val photoUrls: List<String>,
)

@Serializable
data class UserMatchAttendance(
    val id: String,
    @SerialName("user_id") val userId: String,
    @SerialName("match_id") val matchId: String,
    val notes: String? = null,
    @SerialName("photo_urls") val photoUrls: List<String> = emptyList(),
    @SerialName("created_at") val createdAt: String? = null,
)

@Serializable
data class UserMatchAttendanceInsert(
    @SerialName("match_id") val matchId: String,
    @SerialName("user_id") val userId: String,
)

@Serializable
data class AchievementRuleParams(
    val count: Int? = null,
    @SerialName("county_id") val countyId: String? = null,
    @SerialName("sport_code") val sportCode: SportCode? = null,
    val province: Province? = null,
)

@Serializable
data class AchievementDefinition(
    val id: String,
    val code: String,
    val title: String,
    val description: String,
    val icon: String? = null,
    @SerialName("rule_type") val ruleType: String,
    @SerialName("rule_params") val ruleParams: AchievementRuleParams,
    @SerialName("created_at") val createdAt: String? = null,
)

data class HomeAchievementKey(val countyId: String, val sportCode: SportCode)

data class AchievementUnlock(
    val id: String,
    val title: String,
    val description: String,
    val icon: String?,
    val tier: AchievementTier?,
)

enum class AchievementTier {
    STANDARD, BRONZE, SILVER, GOLD;

    val label: String
        get() = when (this) {
            STANDARD -> ""
            BRONZE -> "Bronze"
            SILVER -> "Silver"
            GOLD -> "Gold"
        }

    companion object {
        fun forHomeMatchCount(count: Int): AchievementTier = when {
            count >= 50 -> GOLD
            count >= 25 -> SILVER
            count >= 10 -> BRONZE
            else -> STANDARD
        }
    }
}

data class AchievementProgress(
    val title: String,
    val message: String,
    val icon: String?,
    val tier: AchievementTier,
    val homeGameCount: Int,
)

data class AchievementEvaluation(
    val unlocks: List<AchievementUnlock>,
    val progress: AchievementProgress?,
)

@Serializable
data class UserAchievement(
    val id: String,
    @SerialName("user_id") val userId: String,
    @SerialName("achievement_id") val achievementId: String,
    @SerialName("unlocked_at") val unlockedAt: String,
    val pinned: Boolean = false,
)

@Serializable
data class UserAchievementInsert(
    @SerialName("achievement_id") val achievementId: String,
    @SerialName("user_id") val userId: String,
)

@Serializable
data class UserAchievementPinnedUpdate(
    val pinned: Boolean,
)

@Serializable
data class Friendship(
    val id: String,
    @SerialName("requester_id") val requesterId: String,
    @SerialName("addressee_id") val addresseeId: String,
    val status: String,
    @SerialName("created_at") val createdAt: String? = null,
)

@Serializable
data class FriendshipInsert(
    @SerialName("requester_id") val requesterId: String,
    @SerialName("addressee_id") val addresseeId: String,
)

@Serializable
data class FriendshipStatusUpdate(
    val status: String,
)

@Serializable
data class DevicePushTokenUpsert(
    @SerialName("user_id") val userId: String,
    val token: String,
    @SerialName("updated_at") val updatedAt: String,
)

@Serializable
data class MatchReport(
    val id: String,
    @SerialName("match_id") val matchId: String,
    @SerialName("user_id") val userId: String,
    @SerialName("issue_types") val issueTypes: List<String>,
    val details: String? = null,
    val status: String,
    @SerialName("created_at") val createdAt: String? = null,
)

@Serializable
data class MatchReportInsert(
    @SerialName("match_id") val matchId: String,
    @SerialName("user_id") val userId: String,
    @SerialName("issue_types") val issueTypes: List<String>,
    val details: String? = null,
)
